### PR TITLE
Add RPC call 'getblockstats'

### PR DESCRIPTION
**Pull request by gcomte**
`getblockstats` is available in Bitcoin Core since version `0.17.0` .

> It won't work for some heights with pruning.
> It won't work without -txindex for utxo_size_inc, *fee or *feerate stats.

More info (here)[https://bitcoincore.org/en/doc/0.17.0/rpc/blockchain/getblockstats/]

I'm completely new to Rust and this is my first Rust-PR ever, so please have mercy.

### CODE REVIEW

**Comment by stevenroose** [comment]
It seems that in 0.18 this call needs `-txindex=1`. You can add this to fix that:
```diff
diff --git a/integration_test/run.sh b/integration_test/run.sh
index b40aa78..97925fd 100755
--- a/integration_test/run.sh
+++ b/integration_test/run.sh
@@ -29,6 +29,7 @@ if bitcoind -version | grep -q "v0\.2"; then
 fi
 
 bitcoind -regtest $BLOCKFILTERARG $FALLBACKFEEARG \
+    -txindex=1 \
     -datadir=${TESTDIR}/2 \
     -connect=127.0.0.1:12348 \
     -rpcport=12349 \
```

**Comment by stevenroose** [comment]
The rest looks good to me!

File: client/src/client.rs
Comment (shesek) [inline client/src/client.rs:442]: This can also accept the block hash instead of the height, maybe worth supporting it?

There's also an option for selecting which fields you want statistics for. Though this would require changing all the `GetBlockStatsResult` fields to be an `Option<T>`, which isn't great.

File: json/src/lib.rs
Comment (shesek) [inline json/src/lib.rs:267]: This should be a `bitcoin::Amount` (will also require `with = "bitcoin::util::amount::serde::as_sat"` in the derive).

File: json/src/lib.rs
Comment (shesek) [inline json/src/lib.rs:225]: IIUC there's a preference by the folks here to use `u64` over `u32`/`usize`. Maybe @stevenroose can explain more.

File: json/src/lib.rs
Comment (stevenroose) [reply json/src/lib.rs:267]: Oh yeah, true, we're using those in this crate.

File: client/src/client.rs
Comment (stevenroose) [reply client/src/client.rs:442]: All of them?

File: json/src/lib.rs
Comment (stevenroose) [reply json/src/lib.rs:225]: I suppose this can also be an `Amount`, right?

File: client/src/client.rs
Comment (shesek) [reply client/src/client.rs:442]: If we want to support getting a specific list of fields from `getblockstats`, then I'm afraid yes...

Perhaps it could be a separate struct? So we have `get_block_stats(height) -> GetBlockStatsResult` and `get_block_stats_fields(height, fields) -> GetBlockStatsResultPartial`.

File: json/src/lib.rs
Comment (shesek) [reply json/src/lib.rs:225]: Oh right, this could. But I was referring to all the `u32` fields, not just that one

File: client/src/client.rs
Comment (gcomte) [reply client/src/client.rs:442]: Okay, I proposed some code for this :-) Please check out my last commit @shesek

File: client/src/client.rs
Comment (shesek) [inline client/src/client.rs:446]: Can use `fields: &[bitcoincore_rpc_json::BlockStatsFields]` so it also accepts slices. And `bitcoincore_rpc_json` is also aliased as just `json`.

File: client/src/client.rs
Comment (shesek) [reply client/src/client.rs:442]: Looks good! Commented with two small nits

File: json/src/lib.rs
Comment (gcomte) [reply json/src/lib.rs:267]: As far as I understand, bitcoin::Amount should then also be used for: `avg_fee`, `fee_rate_percentiles`, `max_fee`, `median_fee` and `min_fee`, right?

File: json/src/lib.rs
Comment (gcomte) [reply json/src/lib.rs:225]: Would appreciate some more info about 'using `u64` over `u32`/`usize` _for all fields_'. @stevenroose ?

File: json/src/lib.rs
Comment (gcomte) [reply json/src/lib.rs:267]: When I try too define the fields as `Option<Amount>`, the compiler throws the follwoing error at me: _the trait `SerdeAmount` is not implemented for `std::option::Option<Amount>`_

Any idea how I should deal with this?

Comment (shesek) [reply json/src/lib.rs:267]: Yes, and also `avg_fee_rate`, `max_fee_rate`, `min_fee_rate`, `total_fee`, `subsidy` and `total_out`.

Note that unlike most other RPCs in bitcoind, the feerates here are represented in sat/vB rather than BTC/KB, while amounts are in satoshis rather than BTC.

**Comment by gcomte** [comment]
With commit 03b7fa912a70003ca62d64b631f14e7a243571cc, I have now a version that works (with Amounts), but it has also added a ton of boilerplate code.

Still open is the question, about whether I should use `u64` over `u32` for the fields `avg_tx_size`, `max_tx_sizemax_tx_size`, `median_time`, `median_tx_size`, `min_tx_size`, `time`, and potentially `i64` over `i32` for the fields `utxo_increase` and `utxo_size_inc`.

**Comment by 0xB10C** [comment]
I'd say
- `avg_tx_size`, `max_tx_size`, `median_tx_size` and `min_tx_size` can be a `u32` as single block will currently will always be smaller than 4_000_000 bytes (thus limiting the transaction size to something less)
- `time` and `median_time` should be `u64`
- `utxo_increase` can be `i32` unless I'm missing something, if you really want to be safe then `i64`
- `utxo_size_inc` maybe `i64`, also not sure here

**Comment by gcomte** [comment]
Thanks for your input @0xB10C, I'll wait for more comments from @stevenroose and @shesek about your suggestion.

File: json/src/lib.rs
Comment (stevenroose) [reply json/src/lib.rs]: Just that we use 64-bit numbers for amounts. But in this crate, we are mostly using the dedicated `Amount` type. You can find some other usages of that type and the serde-tag you have to use with it.

**Comment by stevenroose** [comment]
I agree with @0xB10C 's suggestions :)

**Comment by stevenroose** [comment]
Could you also rebase to make CI happy?

**Comment by gcomte** [comment]
Rebased. Travis seems to have a problem still, concerning the `-txindex enabled` on Bitcoin Core versions 18.x .

I've update the data type for timestamps in commit 890c2117faec587a652f15588ea75b27fa050595.
Remind yourselves that this introduces inconsistencies about what data type is used for timestamps within this library: bbf053234f845121e1d16daf31d4122598154434

@0xB10C Here's my rationale for why `i32` should do the job for both `utxo_increase` and `utxo_size_inc`:

Given a maximum block size of `4_000_000` bytes, there is no way how the UTXO set could (`utxo_size_inc`) increase or decrease beyond that number _per-block_ in any way. Consequently, `utxo_increase` will also fit into a `i32`, as it is by definition always closer to `0` than `utxo_size_inc`.

`i32` range: `-2_147_483_648` - `2_147_483_647`

File: integration_test/run.sh
Comment (sgeisler) [inline integration_test/run.sh:12]: The `txindex` argument was added to the wrong `bitcoind` instance imo. There's another one started a few lines below (line 32, idk why), which is used for testing apparently as the `RPC_URL` env var is set to its port. That's probably why the `0.18.*` tests fail, possibly because `-blockfilterindex=1` is set for all other versions and works for stats too?

**Comment by gcomte** [comment]
Just to get a rough idea about the field `utxo_increase` I looked up the past 4500 blocks:

**Lowest increase:**
`for i in $(seq 670000 674578); do bitcoin-cli getblockstats $i | jq -r ".utxo_size_inc"; done | sort -n | head -n 1`
Result: **-732_265**

**Highest increase:**
`for i in $(seq 670000 674578); do bitcoin-cli getblockstats $i | jq -r ".utxo_size_inc"; done | sort -n | tail -n 1`
Result: **1_086_480**

File: integration_test/run.sh
Comment (gcomte) [reply integration_test/run.sh]: You are right, the one below has the flag `-server=1` set, while the one above doesn't.
I moved the flag in commit `f7d1f3bce957be592b6f90d1f7667e2a008a1022`.

**Comment by sgeisler** [comment]
Is this PR still active? Sorry for the wait. We are not good at keeping track of PRs, so pinging reviewers generally helps.

**Comment by gcomte** [comment]
@sgeisler Yes it is still active. I just rebased it, so it should be ready to be merged.

Okay thank you for the information, will ping you guys every now and then :-)

**Comment by shesek** [comment]
All looks good to me, ACK https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/157/commits/6af032a385b096870b06973c8594613f954b7bc8. Thanks for the contribution!

**Comment by gcomte** [comment]
Hey @shesek, good input, thank you!

I tried to address this with the commits https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/157/commits/e0583c4d4376a4a2ec89a2a212c2e40d50bbd66f and https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/157/commits/fdef9f2cc6e445dcbcc09be8d3956ddfe68c92ed.
What do you think?

**Comment by shesek** [comment]
@gcomte Looks good! re-ACK https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/157/commits/fdef9f2cc6e445dcbcc09be8d3956ddfe68c92ed

**Comment by gcomte** [comment]
@sgeisler @stevenroose Looks like this PR is ready to be reviewed/merged.

**Comment by gcomte** [comment]
How is the progress on this guys? Is there something else I could provide you with, to assist you?

**Comment by gcomte** [comment]
I rebased this again, so it is once again, ready to be merged.

**Comment by gcomte** [comment]
I rebased this again, so it is once again, ready to be merged.

**Comment by gcomte** [comment]
@RCasatta thanks for the review, rebased and squashed some commits.

**Review by RCasatta — APPROVED** [review]
utACK 60b0dd768b60ff36bee65b4f017e06812b7ac22d

will merge soon if there is no more feedback

**Comment by gcomte** [comment]
Hey @RCasatta 

Friendly reminder that 3 weeks have passed. Could you specify `soon`?

**Comment by RCasatta** [comment]
Sorry for the delay and thanks for the contribution

**Pull request merged**
